src/Core/MySQL/PacketsProtocolBinary.cpp

```diff
@@ -0,0 +1,363 @@
+                        buffer.write(reinterpret_cast<char *>(&value), 8);
+                        break;
+                    }
+                    case TypeIndex::Float32: {
+                        Float32 value = assert_cast<const ColumnVector<Float32> &>(*col).getData()[row_num];
+                        buffer.write(reinterpret_cast<char *>(&value), 4);
+                        break;
+                    }
+                    case TypeIndex::Float64: {
+                        Float64 value = assert_cast<const ColumnVector<Float64> &>(*col).getData()[row_num];
+                        buffer.write(reinterpret_cast<char *>(&value), 8);
+                        break;
+                    }
+                    case TypeIndex::Date: {
+                        UInt16 value = assert_cast<const ColumnVector<UInt16> &>(*col).getData()[row_num];
+                        LocalDate ld = LocalDate(DayNum(value));
+                        buffer.write(static_cast<char>(4)); // bytes_following
+                        auto year = ld.year();
+                        auto month = ld.month();
+                        auto day = ld.day();
+                        buffer.write(reinterpret_cast<const char *>(&year), 2);
+                        buffer.write(reinterpret_cast<const char *>(&month), 1);
+                        buffer.write(reinterpret_cast<const char *>(&day), 1);
+                        break;
+                    }
+                    case TypeIndex::Date32: {
+                        Int32 value = assert_cast<const ColumnVector<Int32> &>(*col).getData()[row_num];
+                        LocalDate ld = LocalDate(ExtendedDayNum(value));
+                        buffer.write(static_cast<char>(4)); // bytes_following
+                        auto year = ld.year();
+                        auto month = ld.month();
+                        auto day = ld.day();
+                        buffer.write(reinterpret_cast<const char *>(&year), 2);
+                        buffer.write(reinterpret_cast<const char *>(&month), 1);
+                        buffer.write(reinterpret_cast<const char *>(&day), 1);
+                        break;
+                    }
+                    case TypeIndex::DateTime: {
+                        UInt32 value = assert_cast<const ColumnVector<UInt32> &>(*col).getData()[row_num];
+                        String timezone = getDateTimeTimezone(*data_type);
+                        LocalDateTime ldt = LocalDateTime(value, DateLUT::instance(timezone));
+                        int year = ldt.year();
+                        int month = ldt.month();
+                        int day = ldt.day();
+                        int hour = ldt.hour();
+                        int minute = ldt.minute();
+                        int second = ldt.second();
+                        bool has_time = !(hour == 0 && minute == 0 && second == 0);
```

Comment:
That's interesting. DateTimes in ClickHouse are never "timeless", i.e. there is no distinction of values with hour/minute/second = 0 from values where these components are != 0. Does it really make a difference in MySQL?

Reply:
Yes, as it supports packet compression in binary (if I understand the question correctly).

> Type to store a MYSQL_TYPE_DATE, MYSQL_TYPE_DATETIME and MYSQL_TYPE_TIMESTAMP fields in the binary protocol.
>
> To save space the packet can be compressed:
>
> if year, month, day, hour, minutes, seconds and microseconds are all 0, length is 0 and no other field is sent.
> if hour, seconds and microseconds are all 0, length is 4 and no other field is sent.
> if microseconds is 0, length is 7 and micro_seconds is not sent.
> otherwise the length is 11